File: dotnet/src/NetCrunch.Telemetry/Telemetry.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace NetCrunch.Telemetry;

/// <summary>
/// Stages metrics, states and events in memory and flushes them to NetCrunch as a single payload.
/// </summary>
/// <remarks>
/// Instrumentation only mutates memory. A separate flush snapshots the registry and sends absolute
/// current values, so nothing in a request path touches the network, and one request carries every
/// value — which matters because the receiver caps pending payloads per sensor and discards the
/// overflow without reporting it.
/// <para>Instances are safe for concurrent use.</para>
/// <para>See spec/v1.md for the wire format and spec/client-model.md for the behaviour above it.</para>
/// </remarks>
public sealed class Telemetry : IDisposable, IAsyncDisposable
{
    private static readonly JsonSerializerOptions SerializerOptions = new()
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    private readonly object _gate = new();
    private readonly Dictionary<string, Counter> _counters = new(StringComparer.Ordinal);
    private readonly List<Counter> _counterOrder = [];
    private readonly Dictionary<string, StatusEntryDto> _statuses = new(StringComparer.Ordinal);
    private readonly Dictionary<string, Stamp> _timestamps = new(StringComparer.Ordinal);
    private readonly Dictionary<string, Dictionary<string, object?>> _dataObjects = new(StringComparer.Ordinal);
    private readonly List<EventEntryDto> _events = [];

    private readonly SemaphoreSlim _flushGate = new(1, 1);
    private readonly CancellationTokenSource _stopping = new();
    private readonly Task _loop;
    private readonly HttpClient _client;
    private readonly bool _ownsClient;
    private readonly TelemetryOptions _options;

    private int _disposed;

    /// <summary>Creates a registry and, if configured, starts the background flush loop.</summary>
    /// <exception cref="ArgumentException">The endpoint is missing or not an absolute http/https URL.</exception>
    /// <exception cref="ArgumentOutOfRangeException">Retain does not exceed the flush interval.</exception>
    public Telemetry(TelemetryOptions options)
    {
        ArgumentNullException.ThrowIfNull(options);

        if (string.IsNullOrWhiteSpace(options.Endpoint))
        {
            throw new ArgumentException(
                "Endpoint is required — copy it from the Telemetry sensor form.",
                nameof(options));
        }

        if (!Uri.TryCreate(options.Endpoint, UriKind.Absolute, out var uri) ||
            (uri.Scheme != Uri.UriSchemeHttp && uri.Scheme != Uri.UriSchemeHttps))
        {
            throw new ArgumentException("Endpoint must be an absolute http or https URL.", nameof(options));
        }

        if (options.FlushInterval > TimeSpan.Zero && options.Retain <= options.FlushInterval)
        {
            throw new ArgumentOutOfRangeException(
                nameof(options),
                $"Retain ({options.Retain}) must exceed FlushInterval ({options.FlushInterval}), " +
                "or values expire between sends.");
        }

        _options = options;
        _ownsClient = options.HttpClient is null;
        _client = options.HttpClient ?? new HttpClient();

        _loop = options.FlushInterval > TimeSpan.Zero
            ? Task.Run(() => LoopAsync(options.FlushInterval))
            : Task.CompletedTask;
    }

    private readonly record struct Stamp(
        string Object,
        string CounterName,
        string StatusKey,
        DateTimeOffset ObservedAt,
        string StatusValue);

    private async Task LoopAsync(TimeSpan interval)
    {
        using var timer = new PeriodicTimer(interval);
        try
        {
            while (await timer.WaitForNextTickAsync(_stopping.Token).ConfigureAwait(false))
            {
                try
                {
                    await FlushAsync(_stopping.Token).ConfigureAwait(false);
                }
                catch (Exception error) when (error is not OperationCanceledException)
                {
                    _options.OnError?.Invoke(error);
                }
            }
        }
        catch (OperationCanceledException)
        {
            // Disposal.
        }
    }

    // -- staging -----------------------------------------------------------

    /// <summary>
    /// Resolves a counter handle. The same object, counter and instance always return the same
    /// handle, so separate parts of a program instrumenting the same thing converge on one value.
    /// </summary>
    public Counter Counter(string @object, string counterName, string? instance = null)
    {
        Validate.CounterPath(@object, counterName);

        var key = string.Concat(@object, "\0", counterName, "\0", instance ?? string.Empty);

        lock (_gate)
        {
            if (_counters.TryGetValue(key, out var existing))
            {
                return existing;
            }

            var created = new Counter(@object, counterName, instance);
            _counters[key] = created;
            _counterOrder.Add(created);
            return created;
        }
    }

    /// <summary>
    /// Stages a state with an optional explanation. Statuses are what NetCrunch alerting acts on — a
    /// counter on its own raises nothing.
    /// </summary>
    public void Status(string key, string value, string? message = null, bool critical = false, object? data = null)
    {
        Validate.StatusKey(key);
        Validate.StatusValue(value);

        var entry = new StatusEntryDto
        {
            Value = value,
            Message = string.IsNullOrEmpty(message) ? null : message,
            Critical = critical ? true : null,
            Data = data,
        };

        lock (_gate)
        {
            _statuses[key] = entry;
        }
    }

    /// <summary>
    /// Stages a discrete occurrence. Events accumulate and are cleared once sent. Use a status for a
    /// condition that begins and later ends.
    /// </summary>
    public void Event(string message, string? severity = null)
    {
        Validate.EventMessage(message);

        var entry = new EventEntryDto { Message = message, Severity = severity };

        lock (_gate)
        {
            _events.Add(entry);
        }
    }

    /// <summary>Records when something last happened.</summary>
    /// <remarks>
    /// The wire format has no timestamp type, and a raw clock value means nothing outside the process
    /// that produced it. So this becomes two things: an age in seconds, which an alert threshold can
    /// be set on, and a status message carrying the absolute time, for a person to read. The age is
    /// computed at flush time.
    /// </remarks>
    public void Timestamp(
        string @object,
        string counterName,
        string statusKey,
        DateTimeOffset? observedAt = null,
        string statusValue = "OK")
    {
        Validate.CounterPath(@object, counterName);
        Validate.StatusKey(statusKey);
        Validate.StatusValue(statusValue);

        var stamp = new Stamp(@object, counterName, statusKey, observedAt ?? DateTimeOffset.UtcNow, statusValue);

        lock (_gate)
        {
            _timestamps[statusKey] = stamp;
        }
    }

    // -- data objects ------------------------------------------------------

    /// <summary>Stages a data object rendered on the sensor's page.</summary>
    /// <remarks>
    /// The id is the object's identity across payloads: staging the same id again replaces it. There
    /// is no incremental form — a data object is a whole view each time.
    /// </remarks>
    public void Data(string id, DataObject dataObject)
    {
        ArgumentNullException.ThrowIfNull(dataObject);
        Validate.DataObject(id, dataObject.Type, dataObject.Members);

        var encoded = new Dictionary<string, object?>(StringComparer.Ordinal) { ["type"] = dataObject.Type };
        foreach (var member in Validate.DataTypeMembers[dataObject.Type])
        {
            encoded[member] = dataObject.Members[member];
        }

        if (!string.IsNullOrEmpty(dataObject.Name))
        {
            encoded["name"] = dataObject.Name;
        }

        // seriesName labels a plotted series; a table has no series to label.
        if (!string.IsNullOrEmpty(dataObject.SeriesName) && dataObject.Type != "table")
        {
            encoded["seriesName"] = dataObject.SeriesName;
        }

        if (!string.IsNullOrEmpty(dataObject.Message))
        {
            encoded["message"] = dataObject.Message;
        }

        if (!string.IsNullOrEmpty(dataObject.Status))
        {
            encoded["status"] = dataObject.Status;
        }

        lock (_gate)
        {
            _dataObjects[id] = encoded;
        }
    }

    /// <summary>Stages a table.</summary>
    public void Table(string id, TableData table)
    {
        ArgumentNullException.ThrowIfNull(table);
        Data(id, new DataObject
        {
            Type = "table",
            Name = table.Name,
            Message = table.Message,
            Status = table.Status,
            Members = new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["columns"] = table.Columns,
                ["rows"] = table.Rows,
            },
        });
    }

    /// <summary>Stages a time chart. Timestamps are epoch milliseconds.</summary>
    public void TimeSeries(string id, TimeSeriesData series)
    {
        ArgumentNullException.ThrowIfNull(series);
        Data(id, new DataObject
        {
            Type = "time-series",
            Name = series.Name,
            SeriesName = series.SeriesName,
            Message = series.Message,
            Status = series.Status,
            Members = new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["timestamps"] = series.Timestamps,
                ["values"] = series.Values,
            },
        });
    }

    /// <summary>
    /// Stages a labelled bar chart. Named apart from <see cref="Category"/>, which is the
    /// lifetime-bound aggregate — same word in NetCrunch, unrelated meanings.
    /// </summary>
    public void CategoryChart(string id, CategoryChartData chart)
    {
        ArgumentNullException.ThrowIfNull(chart);
        Data(id, new DataObject
        {
            Type = "category",
            Name = chart.Name,
            SeriesName = chart.SeriesName,
            Message = chart.Message,
            Status = chart.Status,
            Members = new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["categories"] = chart.Categories,
                ["values"] = chart.Values,
            },
        });
    }

    // -- lifetime-bound aggregates -----------------------------------------

    /// <summary>Holds one against a counter until disposed.</summary>
    public SelfCount SelfCount(string @object, string counterName, string? instance = null)
        => new(Counter(@object, counterName, instance));

    /// <summary>Contributes a movable amount, withdrawn in full on disposal.</summary>
    public PartCount PartCount(string @object, string counterName, string? instance = null)
        => new(Counter(@object, counterName, instance));

    /// <summary>
    /// Holds one against a single instance at a time, moving it as the value changes. For the chart
    /// of the same name see <see cref="CategoryChart"/>.
    /// </summary>
    public CategoryCount Category(string @object, string counterName)
    {
        Validate.CounterPath(@object, counterName);
        return new CategoryCount(instance => Counter(@object, counterName, instance));
    }

    // -- payload -----------------------------------------------------------

    /// <summary>
    /// Builds the JSON a flush would post, without sending it. Members with nothing in them are
    /// omitted rather than sent empty.
    /// </summary>
    public string BuildPayload(DateTimeOffset? snapshotAt = null)
        => JsonSerializer.Serialize(Snapshot(snapshotAt ?? DateTimeOffset.UtcNow), SerializerOptions);

    private PayloadDto Snapshot(DateTimeOffset snapshotAt)
    {
        lock (_gate)
        {
            var payload = new PayloadDto
            {
                Retain = (int)_options.Retain.TotalMinutes,
                Remove = (int)_options.Remove.TotalMinutes,
            };

            if (_counterOrder.Count > 0 || _timestamps.Count > 0)
            {
                payload.Counters = [];
                foreach (var handle in _counterOrder)
                {
                    payload.Counters.Add(new CounterEntryDto
                    {
                        Path = new CounterPathDto
                        {
                            Object = handle.Object,
                            Counter = handle.CounterName,
                            Instance = handle.Instance,
                        },
                        Value = handle.Value,
                    });
                }
            }

            if (_statuses.Count > 0 || _timestamps.Count > 0)
            {
                payload.Statuses = new Dictionary<string, StatusEntryDto>(_statuses, StringComparer.Ordinal);
            }

            // A timestamp contributes to both collections, so it is expanded here rather than at the
            // call site — the age is only meaningful against this snapshot.
            foreach (var stamp in _timestamps.Values)
            {
                payload.Counters!.Add(new CounterEntryDto
                {
                    Path = new CounterPathDto { Object = stamp.Object, Counter = stamp.CounterName },
                    // Away from zero, so a half second rounds the way every other implementation
                    // rounds it; the framework default is banker's rounding.
                    Value = Math.Round((snapshotAt - stamp.ObservedAt).TotalSeconds, MidpointRounding.AwayFromZero),
                });

                payload.Statuses![stamp.StatusKey] = new StatusEntryDto
                {
                    Value = stamp.StatusValue,
                    Message = stamp.ObservedAt.ToUniversalTime()
                        .ToString("yyyy-MM-dd'T'HH:mm:ss'Z'", CultureInfo.InvariantCulture),
                };
            }

            if (_events.Count > 0)
            {
                payload.Events = [.. _events];
            }

            if (_dataObjects.Count > 0)
            {
                payload.Data = new Dictionary<string, Dictionary<string, object?>>(_dataObjects, StringComparer.Ordinal);
            }

            return payload;
        }
    }

    // -- sending -----------------------------------------------------------

    /// <summary>Posts everything staged as a single request.</summary>
    /// <remarks>
    /// Concurrent calls serialise rather than run together; each sends the absolute state at the
    /// moment it runs. Events are cleared on success. Counters and statuses are kept, so a
    /// long-running process keeps reporting current values without restating them.
    /// </remarks>
    /// <exception cref="TelemetryException">The send failed. The endpoint is never included.</exception>
    public async Task FlushAsync(CancellationToken cancellationToken = default)
    {
        await _flushGate.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            var payload = Snapshot(DateTimeOffset.UtcNow);
            if (payload.IsEmpty)
            {
                return;
            }

            var sentEvents = payload.Events?.Count ?? 0;
            var body = JsonSerializer.Serialize(payload, SerializerOptions);

            await PostAsync(body, cancellationToken).ConfigureAwait(false);

            // Trimmed rather than emptied: events staged while the request was in flight have not
            // been sent, and dropping them would lose them silently.
            lock (_gate)
            {
                _events.RemoveRange(0, Math.Min(sentEvents, _events.Count));
            }
        }
        finally
        {
            _flushGate.Release();
        }
    }

    private async Task PostAsync(string body, CancellationToken cancellationToken)
    {
        TelemetryException? last = null;

        for (var attempt = 1; attempt <= _options.MaxRetries + 1; attempt++)
        {
            var failure = await PostOnceAsync(body, cancellationToken).ConfigureAwait(false);
            if (failure is null)
            {
                return;
            }

            var retryable = failure.StatusCode == 0
                || failure.StatusCode == 429
                || failure.StatusCode >= 500;

            if (!retryable)
            {
                throw failure;
            }

            last = failure;
            if (attempt > _options.MaxRetries)
            {
                break;
            }

            var backoff = TimeSpan.FromSeconds(Math.Min(30, Math.Pow(2, attempt - 1)));
            await Task.Delay(backoff, cancellationToken).ConfigureAwait(false);
        }

        throw last!;
    }

    private async Task<TelemetryException?> PostOnceAsync(string body, CancellationToken cancellationToken)
    {
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeout.CancelAfter(_options.Timeout);

        try
        {
            using var content = new StringContent(body, Encoding.UTF8, "application/json");
            using var response = await _client
                .PostAsync(_options.Endpoint, content, timeout.Token)
                .ConfigureAwait(false);

            if (response.IsSuccessStatusCode)
            {
                return null;
            }

            return new TelemetryException(
                $"NetCrunch telemetry send failed with HTTP {(int)response.StatusCode}.",
                (int)response.StatusCode);
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            throw;
        }
        catch (Exception)
        {
            // Rebuilt, never wrapped: the original names the endpoint, and that URL is currently the
            // credential.
            var reason = timeout.IsCancellationRequested
                ? $"timed out after {_options.Timeout.TotalSeconds:0.#}s"
                : "the endpoint was unreachable";
            return new TelemetryException($"NetCrunch telemetry send failed: {reason}.");
        }
    }

    /// <summary>Discards everything staged.</summary>
    public void Clear()
    {
        lock (_gate)
        {
            _counters.Clear();
            _counterOrder.Clear();
            _statuses.Clear();
            _timestamps.Clear();
            _dataObjects.Clear();
            _events.Clear();
        }
    }

    /// <summary>Stops the flush loop and flushes once more, swallowing any failure.</summary>
    public async ValueTask DisposeAsync()
    {
        if (Interlocked.Exchange(ref _disposed, 1) != 0)
        {
            return;
        }

        await _stopping.CancelAsync().ConfigureAwait(false);
        try
        {
            await _loop.ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
            // Expected.
        }

        try
        {
            await FlushAsync(CancellationToken.None).ConfigureAwait(false);
        }
        catch (TelemetryException error)
        {
            _options.OnError?.Invoke(error);
        }

        Cleanup();
    }

    /// <summary>
    /// Stops the flush loop without a final send. Prefer <see cref="DisposeAsync"/>, which flushes
    /// what is still staged.
    /// </summary>
    public void Dispose()
    {
        if (Interlocked.Exchange(ref _disposed, 1) != 0)
        {
            return;
        }

        _stopping.Cancel();
        Cleanup();
    }

    private void Cleanup()
    {
        _stopping.Dispose();
        _flushGate.Dispose();
        if (_ownsClient)
        {
            _client.Dispose();
        }
    }
}

File: go/telemetry.go
// Package telemetry pushes metrics, states and events from a Go application
// into NetCrunch.
//
// Instrumentation only mutates memory. A separate flush snapshots the registry
// and sends absolute current values, so nothing in a request path touches the
// network, and one request carries every value — which matters because the
// receiver caps pending payloads per sensor and discards the overflow without
// reporting it.
//
// See spec/v1.md for the wire format and spec/client-model.md for the
// behaviour above it.
package telemetry

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// Options configures a Telemetry.
type Options struct {
	// Endpoint is the URL from the Telemetry sensor form. Treat it as a secret:
	// it is never written to an error or a log by this package.
	Endpoint string

	// Token is the bearer token from the Telemetry sensor, sent as an
	// Authorization header. Optional because a sensor need not have one
	// configured, and because servers before NetCrunch 16.0 do not check it;
	// see spec/v1.md section 1.1.
	Token string

	// FlushInterval starts a background flush loop when greater than zero.
	// Close stops it. Zero means flush only when asked.
	FlushInterval time.Duration

	// Retain must exceed FlushInterval, or values expire between sends.
	// Defaults to 5 minutes.
	Retain time.Duration

	// Remove is how long an object survives with no data. Defaults to 24 hours.
	Remove time.Duration

	// Timeout for a single request. Defaults to 30 seconds.
	Timeout time.Duration

	// MaxRetries for transport failures and 5xx responses. Defaults to 3.
	MaxRetries int

	// OnError receives failures from background flushes, which have nowhere
	// else to go. Ignored for explicit Flush calls, which return the error.
	OnError func(error)

	// HTTPClient defaults to a client with no timeout of its own; per-request
	// timeouts come from Timeout.
	HTTPClient *http.Client
}

type statusEntry struct {
	Value    string `json:"value"`
	Message  string `json:"message,omitempty"`
	Critical bool   `json:"critical,omitempty"`
	Data     any    `json:"data,omitempty"`
}

type eventEntry struct {
	Message  string `json:"message"`
	Severity string `json:"severity,omitempty"`
}

type counterPath struct {
	Object   string `json:"object"`
	Counter  string `json:"counter"`
	Instance string `json:"instance,omitempty"`
}

type counterEntry struct {
	Path  counterPath `json:"path"`
	Value float64     `json:"value"`
}

type stampEntry struct {
	object      string
	counter     string
	statusKey   string
	observedAt  time.Time
	statusValue string
}

type payload struct {
	Retain   int                    `json:"retain"`
	Remove   int                    `json:"remove"`
	Counters []counterEntry         `json:"counters,omitempty"`
	Statuses map[string]statusEntry `json:"statuses,omitempty"`
	Events   []eventEntry           `json:"events,omitempty"`
	Data     map[string]any         `json:"data,omitempty"`
}

// Telemetry stages values in memory and flushes them as a single payload.
// It is safe for concurrent use.
type Telemetry struct {
	endpoint   string
	token      string
	retain     time.Duration
	remove     time.Duration
	timeout    time.Duration
	maxRetries int
	onError    func(error)
	client     *http.Client

	mu           sync.Mutex
	counters     map[string]*Counter
	counterOrder []*Counter
	statuses     map[string]statusEntry
	timestamps   map[string]stampEntry
	dataObjects  map[string]map[string]any
	events       []eventEntry

	flushMu sync.Mutex

	stopOnce sync.Once
	stop     chan struct{}
	done     chan struct{}
}

// New validates the options and returns a Telemetry. If FlushInterval is
// greater than zero, a background flush loop starts; Close stops it.
func New(options Options) (*Telemetry, error) {
	if strings.TrimSpace(options.Endpoint) == "" {
		return nil, fmt.Errorf("endpoint is required — copy it from the Telemetry sensor form")
	}
	parsed, err := url.Parse(options.Endpoint)
	if err != nil || !parsed.IsAbs() || (parsed.Scheme != "http" && parsed.Scheme != "https") {
		return nil, fmt.Errorf("endpoint must be an absolute http or https URL")
	}

	retain := options.Retain
	if retain == 0 {
		retain = 5 * time.Minute
	}
	remove := options.Remove
	if remove == 0 {
		remove = 24 * time.Hour
	}
	timeout := options.Timeout
	if timeout == 0 {
		timeout = 30 * time.Second
	}
	maxRetries := options.MaxRetries
	if maxRetries == 0 {
		maxRetries = 3
	}
	client := options.HTTPClient
	if client == nil {
		client = &http.Client{}
	}

	if options.FlushInterval > 0 && retain <= options.FlushInterval {
		return nil, fmt.Errorf(
			"Retain (%s) must exceed FlushInterval (%s), or values expire between sends",
			retain, options.FlushInterval,
		)
	}

	t := &Telemetry{
		endpoint:    options.Endpoint,
		token:       options.Token,
		retain:      retain,
		remove:      remove,
		timeout:     timeout,
		maxRetries:  maxRetries,
		onError:     options.OnError,
		client:      client,
		counters:    make(map[string]*Counter),
		statuses:    make(map[string]statusEntry),
		timestamps:  make(map[string]stampEntry),
		dataObjects: make(map[string]map[string]any),
		stop:        make(chan struct{}),
		done:        make(chan struct{}),
	}

	if options.FlushInterval > 0 {
		go t.loop(options.FlushInterval)
	} else {
		close(t.done)
	}
	return t, nil
}

func (t *Telemetry) loop(interval time.Duration) {
	defer close(t.done)
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	for {
		select {
		case <-t.stop:
			return
		case <-ticker.C:
			if err := t.Flush(context.Background()); err != nil && t.onError != nil {
				t.onError(err)
			}
		}
	}
}

// -- staging ---------------------------------------------------------------

// Counter resolves a counter handle. The same object, counter and instance
// always return the same handle, so separate parts of a program instrumenting
// the same thing converge on one value. Pass "" for instance when there is none.
func (t *Telemetry) Counter(object, counter, instance string) (*Counter, error) {
	if err := validateCounterPath(object, counter); err != nil {
		return nil, err
	}

	key := object + "\x00" + counter + "\x00" + instance

	t.mu.Lock()
	defer t.mu.Unlock()
	if existing, ok := t.counters[key]; ok {
		return existing, nil
	}
	created := &Counter{object: object, counter: counter, instance: instance}
	t.counters[key] = created
	t.counterOrder = append(t.counterOrder, created)
	return created, nil
}

// MustCounter is Counter but panics on an invalid path, for package-level
// declarations where a bad name is a programming error rather than a condition
// to handle.
func (t *Telemetry) MustCounter(object, counter, instance string) *Counter {
	resolved, err := t.Counter(object, counter, instance)
	if err != nil {
		panic(err)
	}
	return resolved
}

// StatusOptions carries the optional parts of a status.
type StatusOptions struct {
	Message  string
	Critical bool
	Data     any
}

// Status stages a state with an optional explanation. Statuses are what
// NetCrunch alerting acts on — a counter on its own raises nothing.
func (t *Telemetry) Status(key, value string, options ...StatusOptions) error {
	if err := validateStatusKey(key); err != nil {
		return err
	}
	if err := validateStatusValue(value); err != nil {
		return err
	}

	entry := statusEntry{Value: value}
	if len(options) > 0 {
		entry.Message = options[0].Message
		entry.Critical = options[0].Critical
		// Guarded rather than assigned straight through: a nil map boxed in an
		// interface is not a nil interface, so `omitempty` would not fire and the
		// payload would carry "data": null for a status that has none.
		if !isNil(options[0].Data) {
			entry.Data = options[0].Data
		}
	}

	t.mu.Lock()
	defer t.mu.Unlock()
	t.statuses[key] = entry
	return nil
}

// Event stages a discrete occurrence. Events accumulate and are cleared once
// sent. Use a status for a condition that begins and later ends.
func (t *Telemetry) Event(message string, severity ...string) error {
	if err := validateEventMessage(message); err != nil {
		return err
	}
	entry := eventEntry{Message: message}
	if len(severity) > 0 {
		entry.Severity = severity[0]
	}

	t.mu.Lock()
	defer t.mu.Unlock()
	t.events = append(t.events, entry)
	return nil
}

// Timestamp records when something last happened.
//
// The wire format has no timestamp type, and a raw clock value means nothing
// outside the process that produced it. So this becomes two things: an age in
// seconds, which an alert threshold can be set on, and a status message carrying
// the absolute time, for a person to read. The age is computed at flush time.
func (t *Telemetry) Timestamp(object, counter, statusKey string, observedAt time.Time) error {
	if err := validateCounterPath(object, counter); err != nil {
		return err
	}
	if err := validateStatusKey(statusKey); err != nil {
		return err
	}

	t.mu.Lock()
	defer t.mu.Unlock()
	t.timestamps[statusKey] = stampEntry{
		object:      object,
		counter:     counter,
		statusKey:   statusKey,
		observedAt:  observedAt,
		statusValue: "OK",
	}
	return nil
}

// -- data objects ----------------------------------------------------------

// DataObject is the generic form of a data object, behind Table, TimeSeries and
// CategoryChart. Members holds the type-specific arrays.
type DataObject struct {
	Type       string
	Name       string
	SeriesName string
	Message    string
	Status     string
	Members    map[string]any
}

// Data stages a data object rendered on the sensor's page.
//
// The id is the object's identity across payloads: staging the same id again
// replaces it. There is no incremental form — a data object is a whole view
// each time.
//
// A data object's Status is part of what is displayed. Alerting acts on
// statuses; a red table is not an alert.
func (t *Telemetry) Data(id string, object DataObject) error {
	if err := validateDataObject(id, object.Type, object.Members); err != nil {
		return err
	}

	encoded := map[string]any{"type": object.Type}
	for _, member := range dataTypeMembers[object.Type] {
		encoded[member] = object.Members[member]
	}
	if object.Name != "" {
		encoded["name"] = object.Name
	}
	// seriesName labels a plotted series; a table has no series to label.
	if object.SeriesName != "" && object.Type != "table" {
		encoded["seriesName"] = object.SeriesName
	}
	if object.Message != "" {
		encoded["message"] = object.Message
	}
	if object.Status != "" {
		encoded["status"] = object.Status
	}

	t.mu.Lock()
	defer t.mu.Unlock()
	t.dataObjects[id] = encoded
	return nil
}

// Table stages a table. Every row must have as many cells as there are columns.
type Table struct {
	Name    string
	Message string
	Status  string
	Columns []any
	Rows    [][]any
}

// Table stages a table data object.
func (t *Telemetry) Table(id string, table Table) error {
	rows := make([]any, len(table.Rows))
	for i, row := range table.Rows {
		rows[i] = row
	}
	return t.Data(id, DataObject{
		Type:    "table",
		Name:    table.Name,
		Message: table.Message,
		Status:  table.Status,
		Members: map[string]any{"columns": table.Columns, "rows": rows},
	})
}

// TimeSeries stages a time chart. Timestamps are epoch milliseconds and must be
// the same length as Values.
type TimeSeries struct {
	Name       string
	SeriesName string
	Message    string
	Status     string
	Timestamps []int64
	Values     []float64
}

// TimeSeries stages a time-series data object.
func (t *Telemetry) TimeSeries(id string, series TimeSeries) error {
	return t.Data(id, DataObject{
		Type:       "time-series",
		Name:       series.Name,
		SeriesName: series.SeriesName,
		Message:    series.Message,
		Status:     series.Status,
		Members:    map[string]any{"timestamps": series.Timestamps, "values": series.Values},
	})
}

// CategoryChart stages a labelled bar chart. Categories and Values must be the
// same length.
type CategoryChart struct {
	Name       string
	SeriesName string
	Message    string
	Status     string
	Categories []string
	Values     []float64
}

// CategoryChart stages a category data object. It is named apart from Category,
// which is the lifetime-bound aggregate — same word in NetCrunch, unrelated
// meanings.
func (t *Telemetry) CategoryChart(id string, chart CategoryChart) error {
	return t.Data(id, DataObject{
		Type:       "category",
		Name:       chart.Name,
		SeriesName: chart.SeriesName,
		Message:    chart.Message,
		Status:     chart.Status,
		Members:    map[string]any{"categories": chart.Categories, "values": chart.Values},
	})
}

// -- lifetime-bound aggregates ---------------------------------------------

// SelfCount holds one against a counter until Close.
//
//	lease := stats.SelfCount("Pool", "Leases Active", "")
//	defer lease.Close()
func (t *Telemetry) SelfCount(object, counter, instance string) *SelfCount {
	handle := t.MustCounter(object, counter, instance)
	handle.Inc()
	return &SelfCount{counter: handle}
}

// PartCount contributes a movable amount, withdrawn in full on Close.
func (t *Telemetry) PartCount(object, counter, instance string) *PartCount {
	return &PartCount{counter: t.MustCounter(object, counter, instance)}
}

// Category holds one against a single instance at a time, moving it as the value
// changes. For the chart of the same name, see CategoryChart.
func (t *Telemetry) Category(object, counter string) *CategoryCount {
	return &CategoryCount{
		resolve: func(instance string) *Counter {
			return t.MustCounter(object, counter, instance)
		},
	}
}

// -- payload ---------------------------------------------------------------

// BuildPayload returns the JSON a flush would post, without sending it. Members
// with nothing in them are omitted rather than sent empty.
func (t *Telemetry) BuildPayload(snapshotAt time.Time) ([]byte, error) {
	return json.Marshal(t.snapshot(snapshotAt))
}

func (t *Telemetry) snapshot(snapshotAt time.Time) payload {
	t.mu.Lock()
	defer t.mu.Unlock()

	built := payload{
		Retain: int(t.retain.Minutes()),
		Remove: int(t.remove.Minutes()),
	}

	for _, handle := range t.counterOrder {
		built.Counters = append(built.Counters, counterEntry{
			Path: counterPath{
				Object:   handle.object,
				Counter:  handle.counter,
				Instance: handle.instance,
			},
			Value: handle.Value(),
		})
	}

	if len(t.statuses) > 0 || len(t.timestamps) > 0 {
		built.Statuses = make(map[string]statusEntry, len(t.statuses)+len(t.timestamps))
		for key, entry := range t.statuses {
			built.Statuses[key] = entry
		}
	}

	// A timestamp contributes to both collections, so it is expanded here rather
	// than at the call site — the age is only meaningful against this snapshot.
	for _, stamp := range t.timestamps {
		built.Counters = append(built.Counters, counterEntry{
			Path:  counterPath{Object: stamp.object, Counter: stamp.counter},
			Value: float64(int64(snapshotAt.Sub(stamp.observedAt).Round(time.Second) / time.Second)),
		})
		built.Statuses[stamp.statusKey] = statusEntry{
			Value:   stamp.statusValue,
			Message: stamp.observedAt.UTC().Format("2006-01-02T15:04:05Z"),
		}
	}

	if len(t.events) > 0 {
		built.Events = append([]eventEntry(nil), t.events...)
	}
	if len(t.dataObjects) > 0 {
		built.Data = make(map[string]any, len(t.dataObjects))
		for id, object := range t.dataObjects {
			built.Data[id] = object
		}
	}
	return built
}

// -- sending ---------------------------------------------------------------

// Flush posts everything staged as a single request.
//
// Concurrent calls serialise rather than run together; each sends the absolute
// state at the moment it runs. Events are cleared on success. Counters and
// statuses are kept, so a long-running process keeps reporting current values
// without restating them.
func (t *Telemetry) Flush(ctx context.Context) error {
	t.flushMu.Lock()
	defer t.flushMu.Unlock()

	built := t.snapshot(time.Now())
	if built.Counters == nil && built.Statuses == nil && built.Events == nil && built.Data == nil {
		return nil
	}

	sentEvents := len(built.Events)

	body, err := json.Marshal(built)
	if err != nil {
		return fmt.Errorf("netcrunch telemetry payload could not be encoded: %w", err)
	}
	if err := t.post(ctx, body); err != nil {
		return err
	}

	// Trimmed rather than emptied: events staged while the request was in flight
	// have not been sent, and dropping them would lose them silently.
	t.mu.Lock()
	t.events = append([]eventEntry(nil), t.events[sentEvents:]...)
	t.mu.Unlock()
	return nil
}

// Clear discards everything staged.
func (t *Telemetry) Clear() {
	t.mu.Lock()
	defer t.mu.Unlock()
	t.counters = make(map[string]*Counter)
	t.counterOrder = nil
	t.statuses = make(map[string]statusEntry)
	t.timestamps = make(map[string]stampEntry)
	t.dataObjects = make(map[string]map[string]any)
	t.events = nil
}

// Close stops the background flush loop, if one is running, and flushes once
// more. Safe to call more than once.
func (t *Telemetry) Close(ctx context.Context) error {
	t.stopOnce.Do(func() { close(t.stop) })
	<-t.done
	return t.Flush(ctx)
}
